### PR TITLE
[AMDGPU] Insert spill codes for the SGPRs used for EXEC copy

### PR DESCRIPTION
**Pull request by cdevadas**
The SGPR registers used for preserving EXEC mask while lowering the whole-wave register spills and copies should be preserved at the prolog and epilog if they are in the CSR range. It isn't happening when there is only wwm-copy lowered and there are no wwm-spills. This patch addresses that problem.

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/SILowerWWMCopies.cpp [hunk 1]
```diff
@@ -137,5 +137,10 @@ bool SILowerWWMCopies::runOnMachineFunction(MachineFunction &MF) {
     }
   }
 
+  // SGPRs reserved for exec copy should be preserved as we encountered WWM_COPY
+  // instances.
+  if (Changed)
+    MFI->setPreserveExecCopyReservedReg();
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SILowerWWMCopies.cpp:143]: Do you really need to track a separate field for this? Can't you just check if RegForExecCopy is used?

File: llvm/lib/Target/AMDGPU/SIFrameLowering.cpp [hunk 2]
```diff
@@ -1515,11 +1517,15 @@ void SIFrameLowering::determinePrologEpilogSGPRSaves(
...
+      assert(!MFI->hasPrologEpilogSGPRSpillEntry(ReservedRegForExecCopy) &&
              "Re-reserving spill slot for EXEC copy register");
-      getVGPRSpillLaneOrTempRegister(MF, LiveUnits, ReservedReg, RC,
+      getVGPRSpillLaneOrTempRegister(MF, LiveUnits, ReservedRegForExecCopy, RC,
                                      /*IncludeScratchCopy=*/false);
     }
+  } else if (ReservedRegForExecCopy) {
+    // Reset it at this point. There are no whole-wave copies and spills
+    // encountered.
+    MFI->setSGPRForEXECCopy(AMDGPU::NoRegister);
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIFrameLowering.cpp:1528]: Does this actually happen? I thought we only reserved this after we know we need it

**Review by arsenm — COMMENTED** [review]
Can you add a mir test?

File: llvm/lib/Target/AMDGPU/SIFrameLowering.cpp (hunk 2 above)
Comment (cdevadas) [reply llvm/lib/Target/AMDGPU/SIFrameLowering.cpp:1528]: Yes. It happens when we have the spills inserted during SILowerSGPRSpills and at that point we ensure there is a reserved register for Exec copy. But vgpr-regalloc didn't insert a liverange spill/copy for the wwm-registers.

**Comment by cdevadas** [comment]
MIR test added.

**Review by arsenm — APPROVED** [review]

**Pull request merged**